src/components/Calendar/indicators.js

```diff
@@ -1,35 +1,52 @@
-import React, { PropTypes } from 'react';
-import { compose } from '../../utils/utils';
+import React, { PropTypes, Component } from 'react';
+import { flowRight as compose, memoize } from 'lodash';
 import Day from '../Day/Day';
-import withProps from '../withProps';
 import { indicatorDay } from '../Day/indicatorDay';
 import { isSameDay } from '../../utils/date-utils';
 
-const dateHasEvent =
-  (date, events) => events.some(
-    (event) => isSameDay(event.date, date)
-  );
-
 /**
- * Higher Order Component to add "day indicators" feature
- * @param {Component} Component
+ * Higher Order Component to add "day indicators" feature to Calendar
+ * @param {Component} CalendarComponent
  * @return {IndicatorCalendar}
  */
-export function indicators(Component) {
-  class IndicatorCalendar extends React.Component {
+export function indicators(CalendarComponent) {
+  class IndicatorCalendar extends Component {
     constructor() {
       super(...arguments);
 
-      this.DayComponent = compose([
-        withProps((props) => ({
-          hasIndicator: dateHasEvent(props.date, this.props.events),
-        })),
+      this.dayHasEvent = this.dayHasEvent.bind(this);
+
+      // This ensures we only enhance once per props.DayComponent
+      // and avoid unnecessarily enhancing on every render
+      this.enhanceDayComponent = memoize(this.enhanceDayComponent.bind(this));
+    }
+
+    dayHasEvent(date) {
+      return this.props.events.some(
+        (event) => isSameDay(event.date, date)
+      );
+    }
+
+    /**
+     * @param {Component} DayComponent
+     * @return {Component} The enhanced DayComponent
+     */
+    enhanceDayComponent(DayComponent) {
+      return compose([
+        // 02 - Provide the enhanced DayComponent with
+        // the required hasIndicator(Date):bool function
+        (DayComponent) => (props) => (
+          <DayComponent {...props} hasIndicator={this.dayHasEvent} />
+        ),
+
+        // 01 - Enhance the DayComponent to enable indicators
         indicatorDay,
-      ])(this.props.DayComponent);
+      ])(DayComponent);
     }
 
     render() {
-      return <Component {...this.props} DayComponent={this.DayComponent} />
+      const enhancedDayComponent = this.enhanceDayComponent(this.props.DayComponent);
```

Comment:
📝 this is not a great idea. See https://twitter.com/dan_abramov/status/820346202424246272

> It’s pretty slow to create new components during rendering. Are you 200% sure that’s the only way?

I am sure there is a better way! 🤔...